packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -105,72 +93,196 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
-        depositsOf[challenger] += slashing;
+        balances[_validator] = balance;
+    }
 
-        emit Slashing(validator, challenger, validatorDeposits);
+    /**
+     * @notice Returns output submitter address.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     *
+     * @return Address of output submitter.
+     */
+    function getOutputSubmitter(uint256 _outputIndex) internal view returns (address) {
+        Types.CheckpointOutput memory output = L2_ORACLE.getL2Output(_outputIndex);
+        return output.submitter;
     }
 
     /**
-     * @notice Returns the deposit balance of validator.
+     * @notice Returns a bonding corresponding to the output index. Reverts if a bonding does not exist.
      *
-     * @param _validator The address of validator
+     * @param _outputIndex Index of L2 checkpoint output.
      *
-     * @return Deposit balance of validator.
+     * @return A bonding data.
      */
-    function balanceOf(address _validator) public view returns (uint256) {
-        return depositsOf[_validator];
+    function getBonding(uint256 _outputIndex) external view returns (Types.Bonding memory) {
+        Types.Bonding storage bonding = bonds[_outputIndex];
+        require(
+            bonding.amount > 0 && bonding.expiresAt > 0,
+            "ValidatorPool: bonding is not exists"
+        );
+        return bonding;
     }
 
-    function isValidator(address _validator) public view returns (bool) {
-        return validators[_validator] || _validator == TRUSTED_VALIDATOR;
+    /**
+     * @notice Returns the balances of given address.
+     *
+     * @param _addr Address of validator.
+     *
+     * @return Balance of given address.
+     */
+    function balanceOf(address _addr) external view returns (uint256) {
+        return balances[_addr];
+    }
+
+    function isValidator(address _addr) external view returns (bool) {
+        uint256 index = validatorIndexes[_addr];
+        return validators[index] == _addr;
     }
 
-    function isValidatorAlive(address _validator) public view returns (bool) {
-        return depositsOf[_validator] > 0 || _validator == TRUSTED_VALIDATOR;
+    function validatorCount() external view returns (uint256) {
+        return validators.length;
+    }
+
+    function nextValidator() public view returns (address) {
+        uint256 nextOutputIndex = L2_ORACLE.nextOutputIndex();
+        if (nextOutputIndex > 0) {
+            Types.CheckpointOutput memory output = L2_ORACLE.getL2Output(nextOutputIndex - 1);
+            uint256 rand = uint256(
+                keccak256(
+                    abi.encodePacked(
+                        validators.length,
+                        output.outputRoot,
+                        output.submitter,
+                        output.timestamp,
+                        output.l2BlockNumber
+                    )
+                )
+            ) % validators.length;
+
+            return validators[rand];
```

Comment:
Great.
However, when using the finalized output root, wouldn't there be an issue if there is no finalized one yet?

Reply:
Could you please tell me when can such case happen?

Reply:
There may not be any finalized output yet. 
For example, there will be no finalized output for 7 days after starting L2. In this case, the outputRoot value will not change, so it will only be randomly generated by values like block number.

Reply:
One way is to have `TRUSTED_VALIDATOR` submit L2 output for the first 7 days after launch.